[captive_portal] [AP] Allow the captive portal to be allways active and restart the AP if needed

RFDarter opened the pull request:
# What does this implement/fix?
Allows the captive portal to be always active.

With this change it is possible to switch between wifi networks without re-flashing the device or powering down the wifi network to force the AP to start.

It is also possible to start the AP by not providing any credentials on the captive portal and saving it.

Also the `web_server` and `captive_portal` can now be used simultaneously.
If the web_server is set in the config the captive portal will be accessible via `http://my-device.local/captive_portal` and the web_server via root `http://my-device.local/` 


FR: https://github.com/esphome/feature-requests/issues/2676
FR https://github.com/esphome/feature-requests/issues/3013
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
esphome:
  name: minimal

esp32:
  board: esp32dev
  framework:
    type: esp-idf

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  ap:
    ap_timeout: 30s

ota:
  platform: esphome

logger:
  level: VERBOSE


captive_portal:
  mode: always_active

web_server:
  version: 3
  local: True

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

RFDarter committed:
init

RFDarter committed:
init

RFDarter committed:
remove captive_portal config option from web_server

RFDarter committed:
Revert "Port wifi_component_esp32_arduino from tcpip_adapter to esp_netif (#6476)"

This reverts commit 0bb2773c646fc3b0d1fe65a73e8e98dd57523e95.

RFDarter committed:
captive portal base path changeable

RFDarter committed:
Merge branch 'captive-portal-save' into web_server-captive-portal

RFDarter committed:
Revert "Revert "Port wifi_component_esp32_arduino from tcpip_adapter to esp_netif (#6476)""

This reverts commit 7ff0c4922af61b39b7c67f8b8735790d531465c6.

RFDarter committed:
Merge branch 'dev' into web_server-captive-portal

RFDarter committed:
handly empty ssid

RFDarter committed:
Merge branch 'dev' into change_captve_portal

RFDarter committed:
Merge branch 'dev' into change_captve_portal

RFDarter committed:
use captive_portal::WEB_SERVER_PORTAL_PATH

RFDarter committed:
fix merge

RFDarter committed:
add Todo

RFDarter committed:
use the handle request functions from captive_portal

RFDarter committed:
file not saved???

RFDarter committed:
unused file

RFDarter committed:
.

RFDarter committed:
..

RFDarter committed:
fix

RFDarter committed:
.

RFDarter committed:
can handle fix

RFDarter committed:
remove double save

RFDarter committed:
comment

RFDarter committed:
use the actual set captive portal path

RFDarter committed:
debug

RFDarter committed:
get it to work

RFDarter committed:
setup ap once and the use start/stop to enable it

RFDarter committed:
fix setup ap once

RFDarter committed:
USE_WIFI_AP wrapper

RFDarter committed:
cleanup

grambas commented:
Was waiting this feature for long time. Thanks for your work. Need more help on testing it on esp32?

RFDarter commented:
> Was waiting this feature for long time. Thanks for your work. Need more help on testing it on esp32?

Testing is always welcome. You should be able to provide this pr as an external component and specify the components`captive_portal`, `web_server` and `wifi`
See: https://esphome.io/components/external_components.html

grambas commented:
> > Was waiting this feature for long time. Thanks for your work. Need more help on testing it on esp32?
> 
> Testing is always welcome. You should be able to provide this pr as an external component and specify the components`captive_portal`, `web_server` and `wifi` See: https://esphome.io/components/external_components.html

So, I tried to compile my ESP32 with your version and encountered an issue. In the wifi/__init__.py file, I had to change all instances of cv.ipv4address to cv.ipv4 between lines 94-107. Maybe the issue is due to a different version of ESPHome? I'm using ESPHome 2024.12.4.

After making the change, I was able to successfully compile and start the ESP32 device. After connecting to the AP, my Android device suggested I go directly to the authorization page for the ESP32. The authorization page opened the web server, but with the active captive portal, shouldn’t it display the exact captive portal page instead?

RFDarter commented:
> Maybe the issue is due to a different version of ESPHome? I'm using ESPHome 2024.12.4.

yep, using the dev version should fix hat.



> The authorization page opened the web server, but with the active captive portal, shouldn’t it display the exact captive portal page instead?

What´s the "exact captive portal page" ?

if you have `web_server` specified in your yaml it should open the webserver.
And under http://<device-name.local>/captive_portal it should display the captive portal.

If `web_server` is not specified in the yaml it should open the captive portal instead.

grambas commented:
there is some mechanics to allow client to detect if network has captive portal or what url is for network landing / authorisation page: https://developer.android.com/about/versions/11/features/captive-portal
android can detect it throw dhcp handshake for example. Im wondering if its configures explicit 192.168.4.1 in esp home and if it makes sense to set it to new captive portal page or leave 192.168.4.1 as now

RFDarter commented:
I think currently the device does not send a captive portal url on the dhcp-handshake.
If that would be the case it would auto open the browser with the given url, or am i mistaken?

What we have on Android on the wifi settings page at the moment is just a link to the gateway give by dhcp.

RFDarter commented:
I just saw this
https://github.com/esphome/esphome/pull/7335
Maybe we could make use of this somehow?

tadeastehan commented:
Hello, I've tried your PR and I am getting this error:
```cpp
Compiling .pioenvs\2-slave-spectral\src\esphome\core\entity_base.cpp.o
src/esphome/components/wifi/wifi_component.cpp: In member function 'virtual void esphome::wifi::WiFiComponent::loop()':
Compiling .pioenvs\2-slave-spectral\src\esphome\core\helpers.cpp.o
src/esphome/components/wifi/wifi_component.cpp:142:12: error: enumeration value 'WIFI_COMPONENT_STATE_START_AP' not handled in switch [-Werror=switch]
  142 |     switch (this->state_) {
      |            ^
Compiling .pioenvs\2-slave-spectral\src\esphome\core\log.cpp.o
cc1plus.exe: some warnings being treated as errors
*** [.pioenvs\2-slave-spectral\src\esphome\components\wifi\wifi_component.cpp.o] Error 1
```
My esphome version: `Version: 2025.3.0-dev` I've already tried changing to a different dev version `2025.2.0b1` but the same error occured. I am using esp32c6 
```yaml
esp32:
  board: seeed_xiao_esp32c6
  variant: ESP32C6
  framework:
    type: esp-idf
    version: 5.3.0
    platform_version: 6.9.0
    sdkconfig_options:
      CONFIG_ESPTOOLPY_FLASHSIZE_4MB: y
```
Could this be the origin of the problem?

RFDarter committed:
encapsulate captive portal

RFDarter commented:
Hi, I removed `WIFI_COMPONENT_STATE_START_AP` a few commits ago since I did not need it.
seems like you compiled it some time before that.

```yaml
esp32:
  board: esp32-c6-devkitc-1
```
fixed it since it had to recompile everything.

a
`esphome clean <yaml>`
on your original yaml and then
`esphome run <yaml>`

should fix it.


So it everything works as expected? It switches AP´s just fine?
It also starts the AP of the esp just fine if you dont provide any SSID on the captive portal?

RFDarter committed:
Merge branch 'dev' into change_captve_portal

tadeastehan commented:
Hello,

I tried removing the ESPHome folder files using `esphome clean <yaml>` as you suggested. It worked. The file compiled correctly and was successfully uploaded to the Seeed Studio ESP32-C6.

I haven’t specified any WiFi configuration in the YAML file, so after 30 seconds, the AP became active. I connected to the AP (hotspot), and both the web server and the captive portal were working.

However, when I used the captive portal to connect the ESP to my WiFi, it successfully connected—I could see the ESP on my local network. But when I visited its IP address, the web server was accessible. However, when I visited `192.168.247.216/captive_portal`, the webpage only displayed "File not found."

The HTML output was:
```html
<body class="volumecontrol-initialized" data-new-gr-c-s-check-loaded="8.927.0" data-gr-ext-installed="" data-gr-ext-disabled="forever">File not found</body>
```

After unplugging the device from power and plugging it back in, the ESP reconnected to the last known WiFi, and the captive portal was available again.

Next, I tested connecting to a different hotspot (from my laptop) using the captive portal. The ESP32 successfully connected to my laptop's hotspot, and the web server and the captive portal were available. Then, I reconnected it to my previous mobile hotspot, which also worked. After that, the captive portal remained available.

So, the issue seems to occur only the first time after using the captive portal—it doesn't work immediately, but after a reboot, everything functions as expected. After trying the sequence for the second time, it worked without rebooting with a physical button, but I could see the device rebooting in the logs. 
***
**Sequence**
Power On → ESP32 making AP (his hotspot) → Connected to `192.168.4.1` web server and captive portal working → Used the captive portal to connect to my mobile phone hotspot → Worked. → Connected to `192.168.247.216` and web server and captive portal working → Used the captive portal to connect to my laptop hotspot → Worked. → Connected to web server and captive portal on laptop hotspot.
***
I have a question: Is it possible to programmatically turn the web server and captive portal on and off?

What I want to achieve is to have the AP and WiFi turned off by default. Then, when the user clicks a button, they should be enabled. I was able to do this using:
```cpp
id(wifi_component).enable();
id(wifi_component).disable();
```
However, I believe this completely disables WiFi, which means that, for example, the [ESPHome API](https://esphome.io/components/api.html) wouldn't work, right?

Is there a way to turn off just the web server and captive portal without disabling WiFi entirely? I want to prevent them from using esp resources when they're not needed.
***
And I have one last question: Is there a way to programmatically erase the saved WiFi credentials from flash memory?

I tried running:
```cpp
id(wifi_component).disable();
id(wifi_component).clear_sta();
```
But after using:
```cpp
id(wifi_component).enable();
```
it still connects to the last known WiFi.

How does the connection process work? If I use the ESP captive portal to save multiple WiFi credentials, will it connect to the last saved network or the strongest available one?
***
If you have any questions, feel free to ask—I'd be happy to help!

tadeastehan commented:
> Hi RFDarter,
> 
> I've tried to install your PR. Unfortunately copilation failed with following errors:
> 
> ```
> INFO ESPHome 2025.7.3
> . . . 
> In file included from src/esphome/components/captive_portal/captive_portal.cpp:1:
> src/esphome/components/captive_portal/captive_portal.h:42:8: error: 'bool esphome::captive_portal::CaptivePortal::canHandle(esphome::web_server_idf::AsyncWebServerRequest*)' marked 'override', but does not override
>    42 |   bool canHandle(AsyncWebServerRequest *request) override {
>       |        ^~~~~~~~~
> src/esphome/components/captive_portal/captive_portal.cpp: In member function 'void esphome::captive_portal::CaptivePortal::handle_captive_portal(esphome::web_server_idf::AsyncWebServerRequest*)':
> src/esphome/components/captive_portal/captive_portal.cpp:14:29: error: 'class esphome::web_server_idf::AsyncWebServerRequest' has no member named 'beginResponse_P'; did you mean 'beginResponse'?
>    14 |   auto *response = request->beginResponse_P(200, "text/html", INDEX_GZ, sizeof(INDEX_GZ));
>       |                             ^~~~~~~~~~~~~~~
>       |                             beginResponse
> src/esphome/components/captive_portal/captive_portal.cpp: In member function 'void esphome::captive_portal::CaptivePortal::start(esphome::web_server_idf::String)':
> src/esphome/components/captive_portal/captive_portal.cpp:69:18: error: 'class esphome::web_server_base::WebServerBase' has no member named 'add_ota_handler'; did you mean 'add_handler'?
>    69 |     this->base_->add_ota_handler();
>       |                  ^~~~~~~~~~~~~~~
>       |                  add_handler
> *** [.pioenvs/esp32test/src/esphome/components/captive_portal/captive_portal.cpp.o] Error 1
> ```
> 
> (I've not copied the messages which are - from my point of view - irrelevant).
> 
> The relevant information from yaml-file look like this:
> 
> ```
> esphome:
>   name: esp32test
>   friendly_name: esp32test
> 
> esp32:
>   board: esp32dev
>   framework:
>     type: esp-idf
> 
> external_components:
>   - source:
>       type: git
>       url: https://github.com/RFDarter/esphome.git
>       ref: change_captve_portal
>     components: [ captive_portal, web_server, wifi ]
> . . .
> ```
> 
> I also tried with board d1_mini_pro (which I'd prefer to use), I tried with ESPHome from HomeAssistant and Standalone-ESPHome on Debian and also after cleaning build files.
> 
> Any clue what went wrong? If I missed to add some relevant information please let me know.
> 
> Thanks Marcus

Hello, try it with the older version of ESPHome 2025.2.0. And if the problem still occurs, I am planning to somehow make a PR with the new version that works with ESPHome 2025.7.0.

RFDarter commented:
Hi, I haven't had time to update this PR.

To have any chance to get merged it still would need docs for it to start the review process.

miikasyvanen commented:
Hi,

Thanks for this PR, just what I needed! I am using 2025.10.0-dev, added changes to files and it compiles and worked like it should :) Now I can use STA and AP modes at the same time to access my custom webserver with or without router and also have captive portal to change credientials if needed. Hope this will some day merged to esphome, gives much more possibilities :)

r-jager-priv commented:
> Hi,
> 
> Thanks for this PR, just what I needed! I am using 2025.10.0-dev, added changes to files and it compiles and worked like it should :) Now I can use STA and AP modes at the same time to access my custom webserver with or without router and also have captive portal to change credientials if needed. Hope this will some day merged to esphome, gives much more possibilities :)

What did you change to get it working under 2025.10.0? I'm on 2025.12.0 after migrating from 2025.03 and would like to get this up and running again.

RFDarter committed:
Merge branch 'dev' into change_captve_portal

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes
